### PR TITLE
mlpack: 1.0.12

### PR DESCRIPTION
**Pull request by saketkc**

### CODE REVIEW

**Comment by dpo** [comment]
Thanks! Will this close #1374?

**Comment by dpo** [comment]
I used to have the same test errors. I think they should be taken upstream.

**Comment by saketkc** [comment]
Apparently, no. 

```
[WARN ] Attempted to save HDF5 data to 'test_file.h5', but Armadillo was compiled without HDF5 support.  Save failed.
[WARN ] Attempted to save HDF5 data to 'test_file.hdf5', but Armadillo was compiled without HDF5 support.  Save failed.
[WARN ] Attempted to save HDF5 data to 'test_file.hdf', but Armadillo was compiled without HDF5 support.  Save failed.
[WARN ] Attempted to save HDF5 data to 'test_file.he5', but Armadillo was compiled without HDF5 support.  Save failed.
```

**Comment by saketkc** [comment]
It's own build system does not report any errors:
http://big.cc.gt.atl.ga.us/job/mlpack%20-%20git%20commit%20test/

**Comment by saketkc** [comment]
I am going to just omit the tests for now. 
https://github.com/mlpack/mlpack/issues/403#issuecomment-74014453

**Comment by saketkc** [comment]
@dpo Good to merge?

File: mlpack.rb [hunk 1]
```diff
@@ -1,85 +1,41 @@
-require "formula"
-
 class Mlpack < Formula
   homepage "http://www.mlpack.org"
-  url "http://www.mlpack.org/files/mlpack-1.0.8.tar.gz"
-  sha1 "f7fce9d37964fb6ede017d29799155f0a08a3e0e"
+  url "https://github.com/mlpack/mlpack/archive/mlpack-1.0.12.tar.gz"
+  sha1 "30ee6fae928cecbcbf3a73e1da30e4c1c952aed0"
+
```
Comment (dpo) [inline mlpack.rb]: Spurious blank line.

File: mlpack.rb [hunk 2]
```diff
@@ -1,85 +1,41 @@
...
-  option "without-check", "Disable build-time tests (not recommended)"
-
-  # fixes test error in an edge case (may not needed in future version)
-  # See https://mailman.cc.gatech.edu/pipermail/mlpack/2014-March/000294.html
-  patch :DATA
+  option "with-check", "Run build-time tests"
 
   def install
     ENV.cxx11 if build.cxx11?
+    dylib = if OS.mac? then "dylib" else "so" end
```
Comment (dpo) [inline mlpack.rb]: Please use ternary operator.

File: mlpack.rb [hunk 3]
```diff
@@ -1,85 +1,41 @@
...
     cmake_args = std_cmake_args
     cmake_args << "-DDEBUG=" + ((build.with? "debug") ? "ON" : "OFF")
     cmake_args << "-DPROFILE=" + ((build.with? "profile") ? "ON" : "OFF")
-    cmake_args << "-DBOOST_ROOT=#{Formula['boost'].opt_prefix}"
-    cmake_args << "-DARMADILLO_INCLUDE_DIR=#{Formula['armadillo'].opt_prefix}/include"
-    cmake_args << "-DARMADILLO_LIBRARY=#{Formula['armadillo'].opt_prefix}/lib/libarmadillo.dylib"
+    cmake_args << "-DBOOST_ROOT=#{Formula['boost'].prefix}"
+    cmake_args << "-DARMADILLO_INCLUDE_DIR=#{Formula['armadillo'].prefix}/include"
+    cmake_args << "-DARMADILLO_LIBRARY=#{Formula['armadillo'].prefix}/lib/libarmadillo.#{dylib}"
```
Comment (dpo) [inline mlpack.rb]: Double quotes here. Why not use opt_prefix?

**Comment by dpo** [comment]
Are you still seeing the test errors? How about with --env=std?

**Comment by saketkc** [comment]
The errors seem to be Mac specific. I have addressed rest of your comments.

**Comment by saketkc** [comment]
@dpo Good to merge?

**Comment by saketkc** [comment]
@dpo

**Comment by dpo** [comment]
Sorry I've been busy. I'm a bit uneasy merging something that has test errors, especially if they are mac specific. At least there should be a clear warning message alerting the user if they enable hdf5 support, don't you think?

**Comment by saketkc** [comment]
I understand. Unfortunately, I do not have access to a Mac to test it out.

**Comment by dpo** [comment]
I'll have a look.

**Comment by saketkc** [comment]
Submitted upstream https://github.com/mlpack/mlpack/issues/417

**Comment by dpo** [comment]
Thanks! I'm observing the same failures and errors related to HDF5.

**Comment by saketkc** [comment]
@BrewTestBot test this please

**Comment by saketkc** [comment]
@BrewTestBot test this please

**Comment by saketkc** [comment]
@dpo Suggestions? Fails on `lion` only. Needs C++11.

**Comment by iml** [comment]
Try `ENV.libcxx if MacOS.version < :mavericks`, that supports C++11 but ML will use stdlibcxx by default.

File: mlpack.rb [hunk 4]
```diff
@@ -1,85 +1,42 @@
-require "formula"
-
 class Mlpack < Formula
   homepage "http://www.mlpack.org"
-  url "http://www.mlpack.org/files/mlpack-1.0.8.tar.gz"
-  sha1 "f7fce9d37964fb6ede017d29799155f0a08a3e0e"
+  # doi "arXiv:1210.6293"
+  #Tagged release post 1.10.12. See: https://github.com/mlpack/mlpack/issues/417#issuecomment-85901766
+  url "https://github.com/mlpack/mlpack/archive/73b48dd.tar.gz"
```
Comment (iml) [inline mlpack.rb]: Can you apply this commit as a patch? The URL should be just the release 1.10.12

Comment (saketkc) [inline mlpack.rb]: I am not really able to patch changes post the tagged release, so took the easy way out.

**Comment by saketkc** [comment]
@BrewTestBot test this please

**Comment by dpo** [comment]
On Mountain Lion:
````
CMake Error at CMakeLists.txt:8 (message):
  No C++11 compiler available!
````
Is this expected?

**Comment by saketkc** [comment]
@BrewTestBot test this please

File: mlpack.rb [hunk 5]
```diff
@@ -1,85 +1,48 @@
...
-  option "without-check", "Disable build-time tests (not recommended)"
+  option "with-check", "Run build-time tests"
 
-  # fixes test error in an edge case (may not needed in future version)
-  # See https://mailman.cc.gatech.edu/pipermail/mlpack/2014-March/000294.html
-  patch :DATA
+  fails_with :clang do
+    build 503
+    cause "clang is not supported. Fails with error: no template named 'forward'"
+  end
```
Comment (iml) [inline mlpack.rb]: Remove this, the build is wrong and it needs to build with clang anyway.

File: mlpack.rb (hunk 4 above)
Comment (iml) [inline mlpack.rb]: Does `https://github.com/mlpack/mlpack/commit/73b48dd.diff` not apply?

File: mlpack.rb [hunk 6]
```diff
@@ -1,85 +1,48 @@
...
+    dylib = OS.mac? ?  "dylib" : "so"
     cmake_args = std_cmake_args
     cmake_args << "-DDEBUG=" + ((build.with? "debug") ? "ON" : "OFF")
     cmake_args << "-DPROFILE=" + ((build.with? "profile") ? "ON" : "OFF")
-    cmake_args << "-DBOOST_ROOT=#{Formula['boost'].opt_prefix}"
-    cmake_args << "-DARMADILLO_INCLUDE_DIR=#{Formula['armadillo'].opt_prefix}/include"
-    cmake_args << "-DARMADILLO_LIBRARY=#{Formula['armadillo'].opt_prefix}/lib/libarmadillo.dylib"
+    cmake_args << "-DBOOST_ROOT=#{Formula["boost"].prefix}"
+    cmake_args << "-DARMADILLO_INCLUDE_DIR=#{Formula["armadillo"].prefix}/include"
+    cmake_args << "-DARMADILLO_LIBRARY=#{Formula["armadillo"].prefix}/lib/libarmadillo.#{dylib}"
```
Comment (iml) [inline mlpack.rb]: Use `opt_prefix`, `opt_include` etc.

**Comment by dpo** [comment]
I still get test failures, but with Armadillo 5.000.1, this builds for me!

**Comment by dpo** [comment]
What's the solution to this on 10.8: `No C++11 compiler available!` ?

**Comment by saketkc** [comment]
Fails on Mountain lion with and without the `fails_with clang` clause.

**Comment by dpo** [comment]
I changed a few things and added more meaningful tests here: https://gist.github.com/fab65ea46e82a52327fc

**Comment by iml** [comment]
ML needs `ENV.cxx11`. Actually `cxx11dep = (build.cxx11?) ? ["c++11"] : []` should really be checking MacOS.version if C++11 is required, and `:cxx11` should not be an option. Anything before Mavericks we need to explicitly ask for C++11 as it is not supported by default with stdlibc++.

**Comment by iml** [comment]
> Fails on Mountain lion with and without the `fails_with clang` clause.

Anything that uses boost, not to mention other dependencies, _must_ be built with clang on the Mac otherwise there will be all sorts of linking errors. See #1593

**Comment by saketkc** [comment]
Closing in lieu of #2145
